encryptor/queryDataEncryptor.go

```diff
@@ -313,6 +316,187 @@ func (encryptor *QueryDataEncryptor) OnQuery(query base.OnQueryObject) (base.OnQ
+	logrus.Debugln("QueryDataEncryptor.encryptInsertValues")
+	tableName := insert.Table.Name
+	// Look for the schema of the table where the INSERT happens.
+	// If we don't have a schema then we don't know what to encrypt, so do nothing.
+	schema := encryptor.schemaStore.GetTableSchema(tableName.String())
+	if schema == nil {
+		logrus.WithField("table", tableName).Debugln("No encryption schema")
+		return values, false, nil
+	}
+
+	// Gather column names from the INSERT query. If there are no columns in the query,
+	// expect a complete list of colums to be available in the schema.
+	var columns []string
+	if len(insert.Columns) > 0 {
+		columns = make([]string, len(insert.Columns))
+		for i, column := range insert.Columns {
+			columns[i] = column.String()
+		}
+	} else if cols := schema.Columns(); len(cols) > 0 {
+		columns = cols
+	}
+
+	changed := false
+	newValues := make([]base.BoundValue, len(values))
+	copy(newValues, values)
+
+	// If there is no column schema available, we can't encrypt values.
+	if len(columns) > 0 {
+		// We can also only process simple queries of the form
+		//
+		//     INSERT INTO table(column...) VALUES ($1, $2, 'static value'...);
+		//
+		// That is, where placeholders uniquely identify the column and used directly
+		// as inserted values of a single row. We don't support functions, casts,
+		// inserting multiple values and query results, etc.
+		//
+		// Walk through the query to find out which placeholders stand for which columns.
+		placeholders := make(map[int]string)
+		switch rows := insert.Rows.(type) {
+		case sqlparser.Values:
+			if len(rows) == 1 {
+				for i, value := range rows[0] {
+					switch value := value.(type) {
+					case *sqlparser.SQLVal:
+						if value.Type == sqlparser.PgPlaceholder {
+							// PostgreSQL placeholders look like "$1". Parse the number out of them.
+							idx, err := strconv.Atoi(strings.TrimPrefix(string(value.Val), "$"))
+							if err != nil {
+								logrus.WithField("placeholder", string(value.Val)).WithError(err).Warning("Invalid placeholder, skipping")
+							}
+							// Placeholders use 1-based indexing and "values" (Go slice) are 0-based.
+							placeholders[idx-1] = columns[i]
+						}
+					}
+				}
+			}
+		}
+
+		// Now that we know the placeholder mapping,
+		// encrypt the values inserted into encrypted columns.
+		for i, columnName := range placeholders {
+			if !schema.NeedToEncrypt(columnName) {
+				continue
+			}
+			settings := schema.GetColumnEncryptionSettings(columnName)
+			switch values[i].Encoding() {
+			case base.BindBinary:
+				changed = true
+				encryptedData, err := encryptor.encryptWithColumnSettings(settings, values[i].Data())
+				if err == ErrUpdateLeaveDataUnchanged {
+					return values, false, nil
+				}
+				newValues[i] = base.NewBoundValue(encryptedData, base.BindBinary)
+			default:
+				// Not supported at the moment.
```

Comment:
for a future, as I understand, we can call `sqlparser.Parse(values[i].Data())` and expect something like `SQLVal` element from parser. Then `switch val.Type` and work with values. If it's hex value or acceptable string, we can encrypt them or tokenize. If it's integer, we can parse it with `strconv.Atoi`, generate new and encode with `strconv.Itoa` to not break format.

Reply:
Mmm... Yeah, that was kinda the plan :+1:

I suspect that "text" will contain literal SQL which can be parsed. It *might* require additional type information though. On the positive side, the encryptor also has a suitable decoder around which might help with byte strings.

However, I did not find a way to trigger this format from tests so I'm not sure what exactly comes here.

Reply:
I've added a TODO into the code to handle this case later. I've also added a warning so that we see it in logs if the text format is being used and we don't handle it.